### ISSUE TITLE
Support for iterating the units for a particular quantity

### ISSUE THREAD

**Issue by bheisler** [iliekturtles/uom#213]
As part of my human-friendly parsing layer (discussed in #212) I would like to be able to construct a set of all valid unit strings for a particular quantity. The `Unit` trait provides convenient access to these strings, but there is no way for me to iterate over the units without manually listing them all in my code.

My use-case for this is that if the user types in "1 pou", I'd like to be able to suggest "Did you mean '1 pound'?".

**Comment by iliekturtles** [comment]
Definitely a good idea! Not sure how `uom` has gone this long without the feature.

What is the best way to implement? Just create a fixed sized static array with all the units? An enum that mirrors the unit types and has some method to conversion from the enum value to the unit value?

**Comment by bheisler** [comment]
I tried to implement this to make a pull request, but I've run into a problem.

Since each unit is a different type, they can't be stored in a container together. The obvious solution is to return `impl Iterator<Item=&'static dyn Unit>` - except that Unit can't be made into a trait object. It can't be made into a trait object because all units implement Copy (system.rs line 163), which is object-unsafe. Additionally, the Unit trait itself is object-unsafe, because the `abbreviation()`, `singular()` and `plural()` functions don't take `&self`. 

Therefore, it is not possible to iterate over multiple different Unit types in any way.

An alternate approach would be to return `impl Iterator<Item=UnitStrings>` where UnitStrings is a separate structure:
```
struct UnitStrings {
    pub abbreviation: &'static str,
    pub singular: &'static str,
    pub plural: &'static str,
}
```

Note that UnitStrings could not provide any connection to its associated unit type, because this would require the UnitStrings struct itself to also be parameterized with the unit type, which means each UnitStrings is a different type.

**Comment by adamreichold** [comment]
> Since each unit is a different type, they can't be stored in a container together.

I am guess it is a bit too much here, but I wanted to mention it since `uom` is already heavy on compile-time logic: We could probably be able to expose the units using [`frunk`](https://crates.io/crates/frunk)'s heterogenous list `HList`? Even though I am not sure whether a stable interface could be provided based on this...

**Comment by bheisler** [comment]
Fixed in #215.
